splunk-otel-android/src/main/java/com/splunk/rum/DiskToZipkinExporter.java

```diff
@@ -0,0 +1,182 @@
+import static com.splunk.rum.SplunkRum.LOG_TAG;
+import static java.util.Collections.emptyList;
+
+import android.util.Log;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import zipkin2.Call;
+import zipkin2.reporter.Sender;
+
+/**
+ * An exporter that pulls pre-encoded zipkin spans from storage and sends them
+ * via a sender. It is bandwidth sensitive and will throttle back if the limit
+ * is exceeded.
+ */
+public class DiskToZipkinExporter {
+
+    private final ScheduledExecutorService threadPool;
+    private final ConnectionUtil connectionUtil;
+    private final Sender sender;
+    private final Path spanFilesPath;
+    private final FileUtils fileUtils;
+    private final BandwidthTracker bandwidthTracker;
+    private final double bandwidthLimit;
+
+    public DiskToZipkinExporter(Builder builder) {
+        this.threadPool = builder.threadPool;
+        this.connectionUtil = builder.connectionUtil;
+        this.sender = builder.sender;
+        this.spanFilesPath = builder.spanFilesPath;
+        this.fileUtils = builder.fileUtils;
+        this.bandwidthTracker = builder.bandwidthTracker;
+        this.bandwidthLimit = builder.bandwidthLimit;
+    }
+
+    void startPolling() {
+        threadPool.scheduleAtFixedRate(this::doExportCycle, 5, 5, TimeUnit.SECONDS);
+    }
+
+    //Visible for testing
+    void doExportCycle() {
+        try {
+            exportPendingFiles();
+        } catch (Exception e) {
+            Log.w(LOG_TAG, "Error processing on-disk files", e);
+        }
+    }
+
+    private void exportPendingFiles() throws IOException {
+        if (!connectionUtil.refreshNetworkStatus().isOnline()) {
+            Log.i(SplunkRum.LOG_TAG, "Network offline, leaving spans on disk for for eventual export.");
+            return;
+        }
+
+        List<Path> pendingFiles = getPendingFiles();
+        boolean sentAnything = false;
+        for (Path file : pendingFiles) {
+
+            double sustainedRate = bandwidthTracker.totalSustainedRate();
+            if (sustainedRate > bandwidthLimit) {
+                Log.i(SplunkRum.LOG_TAG, String.format("Export rate %.2f exceeds limit of %.2f, backing off", sustainedRate, bandwidthLimit));
+                break;
+            }
+
+            boolean dataWasSent = handleFileOnDisk(file);
+            sentAnything |= dataWasSent;
+            if (!dataWasSent) {   // Don't bother trying any remaining files if this one failed.
+                break;
+            }
+        }
+        if (!sentAnything) {
+            bandwidthTracker.tick(emptyList());
+        }
+    }
+
+    private List<Path> getPendingFiles() throws IOException {
+        return fileUtils.listFiles(spanFilesPath)
```

Comment:
The stream returned by `Files.list` must be closed - please use try-with-resources here.

Reply:
The implementation had to change in order to remove usages of `Path`. I don't think it requires closing any longer.

Reply:
No, it looks like it's not needed anymore 👍